### PR TITLE
NIFI-1280 Create FilterCSVColumns Processor.

### PR DESCRIPTION
**Pull request by ToivoAdams**
First version. 
SQL select statement is used to specify how CSV data should be transformed.
Modified Calcite CSV adapter is used for SQL execution.
Some code is borrowed from HiveJdbcCommon, methods convertToCsvStream()

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/resources/TestFilterCSVColumns/US500.csv [hunk 1]
```diff
@@ -0,0 +1 @@
...
"Samira","Heintzman","Mutual Fish Co","8772 Old County Rd #5410","Kent","King","WA",98032,"206-311-4137","206-923-6042","sheintzman@hotmail.com","http://www.mutualfishco.com"
"Margart","Meisel","Yeates, Arthur L Aia","868 State St #38","Cincinnati","Hamilton","OH",45251,"513-617-2362","513-747-9603","margart_meisel@yahoo.com","http://www.yeatesarthurlaia.com"
"Kristofer","Bennick","Logan, Ronald J Esq","772 W River Dr","Bloomington","Monroe","IN",47404,"812-368-1511","812-442-8544","kristofer.bennick@yahoo.com","http://www.loganronaldjesq.com"
"Weldon","Acuff","Advantage Martgage Company","73 W Barstow Ave","Arlington Heights","Cook","IL",60004,"847-353-2156","847-613-5866","wacuff@gmail.com","http://www.advantagemartgagecompany.com"
"Shalon","Shadrick","Germer And Gertz Llp","61047 Mayfield Ave","Brooklyn","Kings","NY",11223,"718-232-2337","718-394-4974","shalon@cox.net","http://www.germerandgertzllp.com"
"Denise","Patak","Spence Law Offices","2139 Santa Rosa Ave","Orlando","Orange","FL",32801,"407-446-4358","407-808-3254","denise@patak.org","http://www.spencelawoffices.com"
"Louvenia","Beech","John Ortiz Nts Therapy Center","598 43rd St","Be
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/resources/TestFilterCSVColumns/US500.csv]: Do we know where this test data came from? We need to be sure to account for licensing information in a header.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        final TestRunner runner = TestRunners.newTestRunner(FilterCSVColumns.class);
+        String sql = "select first_name, last_name, company_name, address, city from CSV.A where city='New York'";
+        runner.setProperty(FilterCSVColumns.SQL_SELECT, sql);
+
+        runner.enqueue(Paths.get("src/test/resources/TestFilterCSVColumns/US500_typeless.csv"));
+        runner.run();
+
+        final List<MockFlowFile> flowFiles = runner.getFlowFilesForRelationship(ExecuteProcess.REL_SUCCESS);
+        for (final MockFlowFile flowFile : flowFiles) {
+            System.out.println(flowFile);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: We should probably be doing some sort of test here, rather than System.out.println to actually verify the results...

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        for (int i = 1; i <= nrofColumns; i++) {
+            System.out.print(resultSet.getMetaData().getColumnLabel(i) + "      ");
+        }
+        System.out.println();
+
+        while (resultSet.next()) {
+            for (int i = 1; i <= nrofColumns; i++) {
+                System.out.print(resultSet.getString(i)+ "  ");
+            }
+            System.out.println();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: We should be using assert's here, to verify the output data

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        int nrofColumns = resultSet.getMetaData().getColumnCount();
+
+        for (int i = 1; i <= nrofColumns; i++) {
+            System.out.print(resultSet.getMetaData().getColumnLabel(i) + "      ");
+        }
+        System.out.println();
+
+        while (resultSet.next()) {
+            for (int i = 1; i <= nrofColumns; i++) {
+                System.out.print(resultSet.getString(i)+ "  ");
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: We should be using assert's here, to verify the output data

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    static protected ResultSet transform(InputStream rawIn, String sql) throws SQLException {
+
+        Reader readerx = new InputStreamReader(rawIn);
+        HashMap<String, Reader> inputs = new HashMap<>();
+        inputs.put("A", readerx);
+
+        Statement statement = null;
+        final Properties properties = new Properties();
+//      properties.setProperty("caseSensitive", "true");
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: We can get rid of this if it's not needed

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+                        convertToCSV(resultSet, out);
+
+                    } catch (final Exception e) {
+                        throw new IOException(e);
+                    }
+                }
+            });
+            session.transfer(transformed, REL_SUCCESS);
+            session.getProvenanceReporter().modifyContent(transformed, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+            logger.info("Transformed {}", new Object[]{original});
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: At this point, we know that the content is of type text/csv so we should probably add that as the mime.type attribute

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        final ProcessorLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        try {
+            FlowFile transformed = session.write(original, new StreamCallback() {
+                @Override
+                public void process(final InputStream rawIn, final OutputStream out) throws IOException {
+                    try (final InputStream in = new BufferedInputStream(rawIn)) {
+
+                        String sql = context.getProperty(SQL_SELECT).getValue();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: It seems like SQL_SELECT should allow the Expression Language to be used. I can certainly envision cases where the user will want to reference a FlowFile Attribute.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        } catch (ProcessException e) {
+            logger.error("Unable to transform {} due to {}", new Object[]{original, e});
+            session.transfer(original, REL_FAILURE);
+        }
+    }
+
+    static protected ResultSet transform(InputStream rawIn, String sql) throws SQLException {
+
+        Reader readerx = new InputStreamReader(rawIn);
+        HashMap<String, Reader> inputs = new HashMap<>();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: This should probably be defined as a Map<String, Reader> rather than HashMap<String, Reader>

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+import org.apache.commons.lang3.time.FastDateFormat;
+
+import au.com.bytecode.opencsv.CSVReader;
+
+
+/** Enumerator that reads from a CSV stream.
+ *
+ * @param <E> Row type
+ */
+class CsvEnumerator2<E> implements Enumerator<E> {
```
Comment (zinking) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java]: why CsvEnumerator2, isn't there a better name?

Comment (joshelser) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java]: Actually, I'm lead to believe (by license and footer) that these were copied (and renamed) from Apache Calcite? Is there something that Calcite could be doing differently that you had to copy (and modify)?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            }
+            System.out.println();
+        }
+    }
+
+    @Test
+    public void testTransformCalc() throws InitializationException, IOException, SQLException {
+        String sql = "select ID, AMOUNT1+AMOUNT2+AMOUNT3 as TOTAL from CSV.A where ID=100";
```
Comment (zinking) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: is there a test case with `stream` in the query, thought that was the whole story.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java (hunk 2 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: Right, I will fix.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java (hunk 3 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: Right, I will fix.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java (hunk 4 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: Right, I will fix.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java (hunk 5 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: Right, I will fix.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java (hunk 6 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: Yes

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java (hunk 7 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: Yes, Expression Language gives flexibility.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java (hunk 8 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: OK

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java (hunk 9 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java]: Yes, It’s modified Calcite CSV adapter.
Current Calcite adapter supports only reading from java.io.File
So CSV adapter is modified to support reading from java.io.InputStream

Modified CSV adapter is currently intermediate code.
In future Calcite should include CSV adapter which is able to read from java.io.InputStream
I (or someone else) will try to create correct CSV adapter.

And name CsvEnumerator2 means just second version of original CsvEnumerator
Please see also Calcite https://mail-archives.apache.org/mod_mbox/calcite-dev/201605.mbox/browser  Reading CSV data from Stream

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java (hunk 10 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: Stream here means reading CSV input from java.io.InputStream
This is not directly related to Calcite streaming queries.
(Select statement includes STREAM keyword)
https://calcite.apache.org/docs/stream.html

Current Calcite adapter supports only reading from java.io.File
So CSV adapter is modified to support reading from java.io.InputStream

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/resources/TestFilterCSVColumns/US500.csv (hunk 1 above)
Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/resources/TestFilterCSVColumns/US500.csv]: https://www.briandunning.com/sample-data/
Free Sample Data for Testing
Can we use this data? Or should we use something else?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java (hunk 10 above)
Comment (zinking) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: what's the big difference between File and InputStream? I thought all conversation is about streaming query.

Comment (ToivoAdams) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestFilterCSVColumns.java]: Yes, reading from File and InputStream and be unified.
But current Calcite CSV adapter supports strictly only File. 

Also CSV adapter supports many ‘tables’. So SQL select supports also joins.
Each ‘table’ is separate File. CSV adapter is capable of discover ‘tables’ using files.
CSV will scan all files in data directory.
Using InputStream requires some kind of solutions for multiple tables.

Some day we might have Calcite CSV adapter which is capable of using both Files and InputStreams.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java (hunk 9 above)
Comment (joshelser) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/calcite/adapter/csv/CsvEnumerator2.java]: > Yes, It’s modified Calcite CSV adapter. Current Calcite adapter supports only reading from java.io.File So CSV adapter is modified to support reading from java.io.InputStream

A name that represents what this does and a class-level javadoc is always great to capture this information :).

I reread Julian's comments, trying to reconcile what he suggested with what you have here. I think the context in which you're using calcite is probably safe (the processor driving things). Neat approach.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        final ProcessorLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        try {
+            FlowFile transformed = session.write(original, new StreamCallback() {
+                @Override
+                public void process(final InputStream rawIn, final OutputStream out) throws IOException {
+                    try (final InputStream in = new BufferedInputStream(rawIn)) {
```
Comment (joshelser) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterCSVColumns.java]: You never user `in` only `rawIn`. This is probably not what you intended (and would be very bad for performance).

**Comment by markap14** [comment]
I don't think we can use the provided test data as-is. Looking at http://www.briandunning.com/sample-data/ I don't see any sort of LICENSE info. Probably best to just mock up a few simple test files manually instead. This will also be easier for testing purposes, I think, so that we can assert specific things in the unit test

**Comment by markap14** [comment]
@ToivoAdams After looking at this a bit more, I've got a few very high-level thoughts. First is that what you have built here is incredibly awesome and powerful! I think FilterCSVColumns is the wrong name for this Processor - it should be perhaps TransformCSV, as leveraging Calcite allows us to do some pretty powerful queries, such as "SELECT * FROM CSV.A WHERE AMOUNT2 < 99" and I imagine that this type of use case will be very common.

I do have a concern, though, which is that if you attempt to perform a JOIN operation, for example "SELECT X.AMOUNT2, X.AMOUNT3 FROM CSV.A as X JOIN CSV.A AS Y ON X.AMOUNT2=Y.AMOUNT2" we end up with an IOException: Stream closed. This is because Calcite will have to read the data multiple times in order to perform the JOIN. I think we can get around this by changing the CsvSchemaFactory2 to be something like CsvInputStreamFactory, and that class, rather than receiving the InputStream directly would be passed the FlowFIle and ProcessSession and could create the InputStream on-demand. This would allow the data to be read multiple times by creating two InputStream's. The nice thing is that if this runs on a system with sufficient RAM the Operating System's disk cache will generally mean that we don't even have to read the data for the second pass unless it's a really massive amount of CSV.

Additionally, I think we need to have some sort of validator for the SQL Select Statement property, as right now if the query is invalid, the processor is valid and just routes everything to failure.

**Comment by markap14** [comment]
@ToivoAdams as I think even more about this, it's opening up a lot of awesome possibilities. I think we could expand this Processor even more into really a "SQLTransform" Processor. I would love to extend this further and even allow other types of data to be evaluated. I'd love to get it into the 0.7.0 baseline. Do you mind if I work on extending what you've provided here to open up some more possibilities?

**Comment by ToivoAdams** [comment]
Go ahead.
Glad to hear you see possibilities.

My full-time job doesn't leave me much time anyway during next few days.
So its actually good to have experienced developer to move thing on.

Other types, you mean Avro, JSON, etc input and output formats?
This means some sort of Calcite adapters must be also developed?

Thanks
toivo

**Comment by markap14** [comment]
Excellent. Yes, I'm thinking potentially Avro, JSON, Parquet... perhaps others. For 0.7.0 we may only have CSV and perhaps JSON but will have things in place to add more. I know a couple of people who I've recently learned are actually Calcite contributors / PMC members who may also be able to provide some guidance to me, so we can make this pretty awesome. You've provided a phenomenally awesome start! It is much appreciated!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 12]
```diff
@@ -235,6 +235,11 @@ language governing permissions and limitations under the License. -->
             <artifactId>json-utils</artifactId>
             <version>0.0.20</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.calcite</groupId>
+            <artifactId>calcite-example-csv</artifactId>
```
Comment (aperepel) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: This seems to be bringing in OpenCSV as a transitive dependency. Between https://commons.apache.org/proper/commons-csv/ and OpenCSV, wouldn't the former be more preferable? Most other contributions seem to be using the commons-csv, few more features, etc. We don't want multiple CSV libraries, do we?

Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: "We don't want multiple CSV libraries, do we?"
I agree it is ideal to avoid multiple libraries that 'do the same thing'.

However...

What we really favor is much more than just that.  Bringing in something like Calcite's libraries means we're adopting their design decisions which they made out of good faith and effort just as we do with internally developed things.  These cases cause natural tension on transitive dependencies.  This is a great example of why we've provided the classloader isolation model we have in NiFi.  https://en.wiktionary.org/wiki/let_a_thousand_flowers_bloom

**Comment by julianhyde** [comment]
@joewitt I agree with your philosophy... however sometimes the other project didn't put as much thought into the decision as you imagine. A little bit of polite back-pressure sometimes helps. If you'd like Calcite to switch from open-csv to commons-csv log a JIRA case and we'd definitely consider it. We needn't halt the Calcite integration for this, but the next version of Calcite might bring in one fewer external dependency.

The same goes for any other dependencies you observe in Calcite.

**Comment by aperepel** [comment]
My only driver was going with something which wasn't abandoned and had recent release activity (who would've thought, that pesky CSV problem... :) )

I was just pointed to this effort this morning, can't yet comment on how portable the OpenCSV use is (please, let it not stop the PR from moving forward - I'd rather refactor in the next release than hold off).

There are other interesting ideas circulating around CSV, which may depend on yet another, third library. For now flagging this with the Calcite team an area to collaborate on is fine, IMHO.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 13]
```diff
@@ -235,6 +235,11 @@ language governing permissions and limitations under the License. -->
...
+            <version>1.7.0</version>
```
Comment (trixpan) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: given months have passed, should we bump this to 1.11.0 ?

Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: @trixpan yes certainly. I took over this ticket but unfortunately haven't had a chance to work on it much until recently. A week or two ago, I picked it up again and plan to have this into the 1.2.0 version of NiFi. I have updated to the latest version in my local branch. Thanks for calling this out!
